Sources/XiEditor/EditView.swift

```diff
@@ -481,9 +483,35 @@ class EditView: NSView, NSTextInputClient, TextPlaneDelegate {
             } else {
                 let builder = TextLineBuilder(line.text, font: font)
                 builder.setFgColor(argb: foregroundArgb)
-                styleMap.applyStyles(builder: builder, styles: line.styles, selColor: selArgb, highlightColors: highlightsArgb)
-                textLine = builder.build(fontCache: renderer.fontCache)
+                styleMap.applyStyles(builder: builder, styles: line.styles)
+
+                // draw annotations
+                for annotationType in AnnotationType.all {
```

Comment:
I'm worried about performance, here; it looks like we're going to be iterating through every annotation for every line.

I'm not sure of the best approach, but my initial sense is that it makes the most sense to treat each annotation type independently; we keep an index into each set of annotations, and for each line we just check if that annotation occurs in this line, advancing the cursors as necessary.

The other thing is that not all annotations are going to be drawn in the same way at the same time. Find and selection are both background colors, but other annotations might need special handling; so I think it makes sense to treat each type uniquely (we could also get a bit fancier and group certain types together; i'm not sure if that would be a win or not).

Reply:
Makes sense.

Reply:
Okay, that is a bit trickier than I thought since annotations of the same type can overlap with each other. So I guess instead of just one index, we'd need to track multiple active indices.

Reply:
hm, not sure I follow. I mean: overlap is possible I guess in some future cases (lints etc) but not in selections or find highlights?

Reply:
For find highlights it is possible (for example if we have the text "testar" and one query for _test_ and one for _star_). And might also be nice to have for future cases.

Reply:
we could then at least keep a single offset which of the place where, for each line, we should start scanning for possible annotations?

Reply:
That is also possible, though maybe it's simpler to just have a set of active annotations that gets then updated for each line. I can just try it out.

Reply:
Keeping track of an offset turned out to be a better solution.